### DISCUSSION TITLE
cq/b123d <-> FreeCAD Path integration

### DISCUSSION THREAD

**Discussion by voneiden** [voneiden/cq-cam#62] [Ideas]
The idea of cq-cam was born out of frustration with wasting unnecessarily time using GUIs.  While the idea is sound and has been proven workable, we are essentially duplicating a lot of effort that has already been done in FreeCAD and the scope is very large for the given resources at hand. 

Meanwhile the Path workbench of FreeCAD is quite capable so I recently got to wonder if it's possible to generate gcode in headless mode and to my surprise it does seem possible programmatically via just the python interface (FreeCAD GUI does not need to be running!). 

I believe this idea might be worth exploring further considering all the features FreeCAD has. Essentially the architecture would be like

`CQ/B123d API --> Preprocessing geometry --> BREP transfer of relevant geometry to a FreeCAD document  --> Configure job, ops, tools, postprocessor in FreeCAD -> Generate G-code in FreeCAD -> Generating a suitable visualization of the G-code for CQ-Editor/ocp-viewer`

**Comment by voneiden** [comment]
Great progress on the PoC!  

![image](https://github.com/voneiden/cq-cam/assets/437576/71fbca36-8240-41f9-81c7-83ef4c3da44b)

**Comment by voneiden** [comment]
Working on this in a separate repo

https://github.com/voneiden/ocp-freecad-cam 

FYI @giannissc

**Comment by giannissc** [comment]
Nice! Is there anything that I can do to help?

**Reply by voneiden** [reply]
The answer is probably, but I'm not sure I can point out anything specific yet. There's plenty of work to do in bringing the rest of the ops (currently got Profile, Face, Pocket) and their configurable properties. Rest of the default tool shapes (currently got Endmill and Ballnose). Tests, documentation, polishing the user facing API.. 

In any case, I've not hit any technical blockers so this really looks like it'll work. It's a bit shame considering the time and effort poured in cq-cam. But it's not all in vain - some code, ideas, etc. have been reused.

**Comment by voneiden** [comment]
![image](https://github.com/voneiden/cq-cam/assets/437576/30c00a04-ec34-4bef-a7d1-f477ead9fb54)

V-Carving faces, wee bit unstable, but it works :sun_with_face: 

I spent quite some time with #25 trying to get this feature to cq-cam without success.

**Comment by giannissc** [comment]
FreeCAD discussed on their recent hackathon in Vancouver about a new backend for the Path Workbench called CavalierContours. They chose not to integrate with it in the end but I took a look at it and it the most complete library available. I will at some point give this another go and either integrate with `cq-cam` or create a new `ocp-cavalier-cam` repo for it

[C++ Library](https://github.com/jbuckmccready/CavalierContours)
[Rust Library](https://github.com/jbuckmccready/cavalier_contours)

**Reply by giannissc** [reply]
What do you think? @voneiden

**Reply by voneiden** [reply]
Hey sorry, I forgot to reply @giannissc . Definitely interesting stuff. Feel free to fork/borrow whatever you want from cq-cam and/or ocp-freecad-cam. If it seems like you'd like to continue with cq-cam just let me know, I can add you as a maintainer.
